pkg/image/registry/imagestreamtag/clone.go

```diff
@@ -0,0 +1,794 @@
+	}
+	if imageClone.Namespace != namespace {
+		return nil, errors.NewBadRequest(fmt.Sprintf("The namespace field on the provided ImageStreamTagClone must match the namespace in the URL."))
+	}
+
+	part, err = mr.NextPart()
+	if err == io.EOF {
+		return nil, errors.NewBadRequest(fmt.Sprintf("You must provide two parts - an application/json or application/vnd.kubernetes.protobuf encoded ImageStreamTagClone object and the binary contents of the layer."))
+	}
+	glog.Infof("Part 2: %v", err)
+	if err != nil {
+		return nil, errors.NewBadRequest(err.Error())
+	}
+
+	if part.Header.Get("Content-Type") != manifest.DockerV2Schema2LayerMediaType {
+		// TODO: in the future support application/zip?
+		err := errors.NewGenericServerResponse(http.StatusNotAcceptable, "post", h.r.rest.gr, h.name, "", 0, false)
+		err.ErrStatus.Message = fmt.Sprintf("The only accepted media type for the layer part is a valid Docker layer (filesystem diff tar+gz) '%s'", manifest.DockerV2Schema2LayerMediaType)
+		return nil, err
+	}
+
+	target, tag, err := imageStreamForClone(h.r.rest.imageStreamRegistry, h.ctx, h.name, h.r.rest.gr, nil)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return nil, err
+		}
+
+		name, _, ok := imageapi.SplitImageStreamTag(h.name)
+		if !ok {
+			return nil, errors.NewBadRequest(fmt.Sprintf("The name field on the provided ImageStreamTagClone is not a valid image stream tag name."))
+		}
+
+		// try to create the target if it doesn't exist
+		target = &imageapi.ImageStream{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: namespace,
+			},
+		}
+	}
+
+	// if we specified a precondition, we must find a match.
+	if len(h.options.PreconditionUID) > 0 && h.options.PreconditionUID != imageClone.UID {
+		return nil, errors.NewConflict(h.r.rest.gr, h.name, fmt.Errorf("the precondition UID does not match the image stream tag clone UID"))
+	}
+
+	return h.r.rest.completeClone(h.ctx, tag, target, toImageStreamTagClone(h.name, imageClone, target), part, contentType)
+}
+
+func uploadLayer(ctx apirequest.Context, r io.Reader, repo distribution.Repository, mediaType string) (distribution.Descriptor, digest.Digest, error) {
+	blobs := repo.Blobs(ctx)
+	bw, err := blobs.Create(ctx)
+	if err != nil {
+		return distribution.Descriptor{}, "", fmt.Errorf("unable to begin uploading blob: %v", err)
+	}
+	pr, pw := io.Pipe()
+
+	// calculate the diffID as the sha256 sum of the layer contents
+	layerhash := sha256.New()
+	ch := make(chan error)
+	go func() {
+		defer close(ch)
+		gr, err := gzip.NewReader(pr)
+		if err != nil {
+			ch <- fmt.Errorf("unable to create gzip reader layer upload: %v", err)
+			return
+		}
+		_, err = io.Copy(layerhash, gr)
+		ch <- err
+	}()
+
+	// calculate the blob digest as the sha256 sum of the uploaded contents
+	blobhash := sha256.New()
+	w := io.MultiWriter(bw, blobhash, pw)
+
+	// upload the bytes
+	n, err := io.Copy(w, r)
```

Comment:
As we discussed below: `io.Copy(bw, io.TeeReader(r, io.MultiWriter(blobhash, pw))` or `bw.ReadFrom(io.TeeReader(r, io.MultiWriter(blobhash, pw)))`